File: common/src/main/java/com/ssblur/scriptor/word/PartialSpell.java
package com.ssblur.scriptor.word;

import com.ssblur.scriptor.api.word.Action;
import com.ssblur.scriptor.api.word.Descriptor;

import java.util.ArrayList;

public record PartialSpell(
  Action action,
  Descriptor... descriptors
) {
  public Descriptor[] deduplicatedDescriptors() {
    ArrayList<Descriptor> out = new ArrayList<>();
    for(var descriptor: descriptors) {
      if(descriptor.allowsDuplicates() || !out.contains(descriptor))
        out.add(descriptor);
    }
    return out.toArray(Descriptor[]::new);
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/word/PartialSpell.kt
package com.ssblur.scriptor.word

import com.ssblur.scriptor.api.word.Action
import com.ssblur.scriptor.api.word.Descriptor

class PartialSpell(val action: Action, vararg val descriptors: Descriptor) {
  fun deduplicatedDescriptors(): Array<Descriptor> {
    val out = ArrayList<Descriptor>()
    for (descriptor in descriptors) {
      if (descriptor.allowsDuplicates() || !out.contains(descriptor)) out.add(descriptor)
    }
    return out.toTypedArray()
  }
}
